### PR TITLE
Rework code of glue layer to native code to allow decoupling SPI drivers

### PR DESCRIPTION
**Pull request by AdrianSoundy**
## Description
Updated SPi assembly to go with nanoCLR changes to separate SPI into a nanoCLR part and Windows.Device.SPI so that the  nanoCLR spi can be shared with other SPI drivers like graphics, ethernet

These changes move device checking/in use into nanoClr 

Backed by nanoframework/nf-interpreter#1637.

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Signed-off-by: adriansoundy <adriansoundy@gmail.com>

### CODE REVIEW

File: source/Windows.Devices.Spi/SpiDevice.cs [hunk 1]
```diff
@@ -39,40 +36,40 @@ internal SpiDevice(string spiBus, Spi​Connection​Settings settings)
...
-                NativeInit();
-
-                // ... and add this device
-                controller.DeviceCollection.Add(this);
-
-                _syncLock = new object();
+                _deviceId = NativeOpenDevice(controllerId);
             }
-            else
+            catch(NotSupportedException ex)
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply source/Windows.Devices.Spi/SpiDevice.cs]: Its in the nanoSPI_OpenDeviceEx() method in nanoHAL_Spi.cpp line 251
```
            // Return NOT_SUPPORTED when Device already in use. Not really any other relevant exception that's
            // currently raised in managed code
            return CLR_E_NOT_SUPPORTED;
```

[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply source/Windows.Devices.Spi/SpiDevice.cs]: Added XML docs for exceptions on FromID() & GetDevice()

File: source/Windows.Devices.Spi/SpiDevice.cs [hunk 2]
```diff
@@ -39,40 +36,40 @@ internal SpiDevice(string spiBus, Spi​Connection​Settings settings)
...
-                _deviceId = deviceId;
-
-                // call native init to allow HAL/PAL inits related with Spi hardware
-                NativeInit();
-
-                // ... and add this device
-                controller.DeviceCollection.Add(this);
-
-                _syncLock = new object();
+                _deviceId = NativeOpenDevice(controllerId);
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply source/Windows.Devices.Spi/SpiDevice.cs]: Done

File: source/Windows.Devices.Spi/SpiBusInfo.cs [hunk 3]
```diff
@@ -50,10 +49,9 @@ public int ChipSelectLineCount
         /// <value>
         /// The clock cycle in Hz.
         /// </value>
-        public extern int MinClockFrequency
+        public int MinClockFrequency
         {
-            [MethodImpl(MethodImplOptions.InternalCall)]
-            get;
+            get { return NativeMinClockFrequency(_controllerId); }
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply source/Windows.Devices.Spi/SpiBusInfo.cs]: Done

File: source/Windows.Devices.Spi/SpiBusInfo.cs [hunk 4]
```diff
@@ -38,10 +38,9 @@ public int ChipSelectLineCount
         /// <value>
         /// The clock cycle in Hz.
         /// </value>
-        public extern int MaxClockFrequency
+        public int MaxClockFrequency
         {
-            [MethodImpl(MethodImplOptions.InternalCall)]
-            get;
+            get { return NativeMaxClockFrequency(_controllerId); }
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply source/Windows.Devices.Spi/SpiBusInfo.cs]: Done

File: source/Windows.Devices.Spi/SpiBusInfo.cs [hunk 5]
```diff
@@ -28,8 +29,7 @@ internal SpiBusInfo(string spiBus)
         /// </value>
         public int ChipSelectLineCount
         {
-            get { return _chipSelectLineCount; }
-            set { _chipSelectLineCount = value; }
+            get { return NativeChipSelectLineCount(_controllerId); }
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply source/Windows.Devices.Spi/SpiBusInfo.cs]: Done

**Comment by AdrianSoundy** [comment]
Should be complete now

**Review by josesimoes — APPROVED** [review]
LGTM

**Pull request merged**